docs/userguide/joins.rst

```diff
@@ -0,0 +1,260 @@
+    sensor_futures = []
+    for sensor in future.result()
+      sensor_futures.append(process_sensor(sensor)
+    assembled_future = assemble_sensors(*sensor_futures)
+    post_process_futures.append(post_process(assembled_future))
+
+  mosaic_future = make_mosaic(*post_process_futures)
+  mosaic_future.result()
+
+Weaknesses
+""""""""""
+
+There is some loss of the sense of data flow expressed in syntax, obscured by
+the implementation of a rudimentary in-workflow task scheduler that only knows about sensor list futures.
+
+This is *still* blocking on the second for loop needing to complete before later apps can be launched,
+and that for loop only completes when all sensor lists have been generated - this is earlier than
+in previous examples, but there is still blocking there.
+
+This impedes compositionality: this code could not be placed into a function and (for example) run
+inside another for loop that loops over multiple datasets: the function would block each time waiting
+for sensor list generation, rather than processing each dataset's sensor list generations
+concurrently.
+
+Anything that blocks the execution thread on future completion (for example, ``.result()``
+or ``.as_completed()``) is the enemy.
+
+`join_app` syntax
+------------------
+
+This is an attempt to move some of the cases where blocking and ad-hoc task scheduling happens in
+the workflow into parsl dependency handling. The only blocking should happen at the very end of the
+workflow, so that the main process does not end until work is completed. Other than that, nothing else
+in the user workflow should block waiting for app completion.
+
+.. code-block:: python
+
+  post_process_futures=[]
+  for patch in patch_list:
+    sensor_list_future = generate_sensor_list_app(patch)
+    sensors_future = process_sensors(sensor_list_future)
+    assembled_future = assemble_sensors(sensors_future)
+    post_process_futures.append(post_process(assembled_future))
+
+  mosaic_future = make_mosaic(*post_process_futures)
+  mosaic_future.result()
+
+  @join_app
+  def process_sensors(sensor_list):
+    sensor_futures = []
+    for sensor in sensor_list:
+      sensor_futures.append(process_sensor(sensor))
+    return combine(*sensor_futures)
+
+  @python_app
+  def combine(*args):
+    pass # do nothing, but only after all args are complete
+
+
+
+This example uses a helper app called ``combine`` which, given a list of input futures,
+completes when all of those futures complete, without any further processing. This constructs a
+barrier future, depending on an arbitrary list of other futures.
+
+This allows more naunced dependencies to be expressed that can help with:
+
+* increased concurrency - helping with strong scaling
+* more focused error propagation - allowing more of an ultimately failing workflow to complete
+* more useful monitoring information
+
+Terminology
+-----------
+
+The term "join" comes from use of monads in functional programming, especially Haskell.
+
```

Comment:
what would happen if we replaced all python_apps by join_apps?  (Or simply changed the implementation of python_app to the implementation of join_app?